### PR TITLE
Multi-Step Entropy Regularization for SAC

### PR DESCRIPTION
**Pull request by frederikschubert**
- Add `record_extra_info` flag to the `NStep` tracer that records the intermediate states in the new `extra_info` field to `TransitionBatch`
- Add support for the `NStepEntropyRegularizer` in `SoftPG`

This PR contains an initial working implementation of the mechanism and sums um the discounted entropy bonuses of the states `s_t`, `s_{t + 1}`, ... , `s_{t + n - 1}` for the soft policy gradient regularization.

### CODE REVIEW

File: coax/reward_tracing/_nstep_verbose.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        states = [s]
+        dones = [False]
+        i = 0
+        while i < self.n and i < len(self):
+            s_next, *_ = self._deque_s[i]
+            states.append(s_next)
+            dones.append(False)
+            i += 1
+        while len(states) != self.n + 1:
+            states.append(s)
```
Comment (frederikschubert) [inline coax/reward_tracing/_nstep_verbose.py:58]: I just noticed that this leads to the entropy of the t+n'th state being added to the regularization which is false. I will change this.

File: coax/policy_objectives/_soft_pg.py [hunk 2]
```diff
@@ -54,4 +54,9 @@ def objective_func(self, params, state, hyperparams, rng, transition_batch, Adv)
         chex.assert_rank([W, Q], 1)
         objective = W * Q
 
+        dist_params, _ = zip(*[self.pi.function(params, state, next(rngs),
+                                                self.pi.observation_preprocessor(
+            next(rngs), s_next), True)
+            for s_next in transition_batch.extra_info['states']])
+        dist_params = (dist_params, jnp.asarray(transition_batch.extra_info['dones']))
```
Comment (frederikschubert) [inline coax/policy_objectives/_soft_pg.py:61]: This does not yet check, whether the extra_info is even there. But I wanted to write a proof of concept before polishing the code.

File: coax/reward_tracing/_nstep_verbose.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+from .._base.errors import InsufficientCacheError
+from ._nstep import NStep
+from ._transition import TransitionBatch
+
+__all__ = (
+    'NStep',
+)
+
+
+class NStepVerbose(NStep):
```
Comment (frederikschubert) [inline coax/reward_tracing/_nstep_verbose.py:34]: We could integrate this into the normal `NStep` tracer but I am not sure whether that's the way to go. We would need to add an argmument to it.

File: coax/reward_tracing/_nstep_verbose.py (hunk 1 above)
Comment (KristianHolsheimer) [reply coax/reward_tracing/_nstep_verbose.py:58]: Could you use the `dones` to mask those contributions?

File: coax/reward_tracing/_nstep_verbose.py (hunk 3 above)
Comment (KristianHolsheimer) [reply coax/reward_tracing/_nstep_verbose.py:34]: Yeah maybe in this case it might be better to add an argument to the existing `NStep` class.

Also, please could you add the intermediate actions and rewards in there too? This would make v-trace easier in the future.

File: coax/reward_tracing/_nstep_verbose.py (hunk 1 above)
Comment (frederikschubert) [reply coax/reward_tracing/_nstep_verbose.py]: Yes, I added some code to filter those!

File: coax/reward_tracing/_nstep_verbose.py (hunk 3 above)
Comment (frederikschubert) [reply coax/reward_tracing/_nstep_verbose.py]: I moved the `extra_info` part to the `NStep` tracer. It is now recording all possible data for future use.

**Comment by frederikschubert** [comment]
This still needs some tests and I will try to do a rudimentary performance comparison.

**Comment by frederikschubert** [comment]
Episode Return SAC vs. TD3

I just ran a quick test on `Pendulum-v1` with TD3 (turquoise) and SAC (red), both with 5-step TD.

Additionally, I ran SAC with `NStepEntropyRegularizer` (blue, red) and `EntropyRegularizer` (turquoise, pink) with 5-step TD and regularization weight of 0.2 (default in many SAC implementations) and 0.2/5=0.04.
Entropy

At least after this one sample run per variant, the performance of the n-step entropy regularization seems to be better and the entropies are also behaving as expected, i.e. more steps in the entropy bonus leads to higher entropy. 

Episode Return SAC

**Comment by KristianHolsheimer** [comment]
> [...] more steps in the entropy bonus leads to higher entropy.

This is interesting. Theoretically the agents should optimize the same objective (given the same hparams), at least asymptotically. I would expect them to converge to roughly the same points (in terms of entropy etc). In other words, I would expect them to differ only in how fast they converge.

File: coax/policy_objectives/_soft_pg.py [hunk 4]
```diff
@@ -54,4 +55,15 @@ def objective_func(self, params, state, hyperparams, rng, transition_batch, Adv)
         chex.assert_rank([W, Q], 1)
         objective = W * Q
 
+        if type(self.regularizer) == NStepEntropyRegularizer:
+            if type(transition_batch.extra_info) != dict:
```
Comment (KristianHolsheimer) [inline coax/policy_objectives/_soft_pg.py:59]: Please could you use `isinstance` for type checking? See e.g. [this post](https://stackoverflow.com/questions/1549801/what-are-the-differences-between-type-and-isinstance).

**Review by KristianHolsheimer — CHANGES_REQUESTED** [review]
Thanks for this.. well written! And thanks for including the tests for `include_extra_info=True` as well.

**Review by KristianHolsheimer — APPROVED** [review]
